tracing-modality/src/layer.rs

```diff
@@ -0,0 +1,323 @@
+    pub fn init() -> Result<Self, InitError> {
+        Self::init_with_options(Default::default())
+    }
+
+    /// Initialize a new `ModalityLayer`, with specified options.
+    pub fn init_with_options(mut opts: Options) -> Result<Self, InitError> {
+        let run_id = Uuid::new_v4();
+        opts.add_metadata("run_id", run_id.to_string());
+
+        let ingest = ModalityIngest::connect(opts).context("connect to modality")?;
+        let ingest_handle = ingest.spawn_thread();
+        let sender = ingest_handle.ingest_sender.clone();
+
+        Ok(ModalityLayer {
+            ingest_handle: Some(ingest_handle),
+            sender,
+        })
+    }
+
+    /// Convert this `Layer` into a `Subscriber`by by layering it on a new instace of `tracing`'s
+    /// `Registry`.
+    pub fn into_subscriber(self) -> impl Subscriber {
+        Registry::default().with(self)
+    }
+
+    /// Take the handle to this layer's ingest instance. This can only be taken once.
+    ///
+    /// This handle is primarily for calling [`ModalityIngestHandle::finish()`] at the end of your
+    /// main thread.
+    pub fn take_handle(&mut self) -> Option<ModalityIngestHandle> {
+        self.ingest_handle.take()
+    }
+
+    fn handle_message(&self, message: ingest::Message) {
+        self.ensure_timeline_has_been_initialized();
+        let wrapped_message = ingest::WrappedMessage {
+            message,
+            tick: START.elapsed(),
+            timeline: Self::LOCAL_METADATA.with(|m| m.thread_timeline),
+        };
+
+        if let Err(_e) = self.sender.send(wrapped_message) {
+            let should_warn = WARN_LATCH
+                .compare_exchange(false, true, Ordering::Relaxed, Ordering::Relaxed)
+                .is_ok();
+
+            if should_warn {
+                eprintln!(
+                    "warning: attempted trace after tracing modality has stopped accepting \
+                     messages, ensure spans from all threads have closed before calling \
+                     `finish()`"
+                );
+            }
+        }
+    }
+
+    fn get_next_span_id(&self) -> LocalSpanId {
+        loop {
+            // ordering of IDs doesn't matter, only uniqueness, use relaxed ordering
+            let id = NEXT_SPAN_ID.fetch_add(1, Ordering::Relaxed);
+            if let Some(id) = NonZeroU64::new(id) {
+                return LocalSpanId(id);
+            }
+        }
+    }
+
+    fn ensure_timeline_has_been_initialized(&self) {
+        if !Self::THREAD_TIMELINE_INITIALIZED.with(|i| i.get()) {
+            Self::THREAD_TIMELINE_INITIALIZED.with(|i| i.set(true));
+
+            let cur = thread::current();
+            let name = cur
+                .name()
+                .map(Into::into)
+                .unwrap_or_else(|| format!("thread-{:?}", cur.id()));
+            let message = ingest::Message::NewTimeline { name };
+            self.handle_message(message)
```

Comment:
Fair enough. I was figuring that the only possible failure mode for misusing this would be an infinite recursion, which any amount of testing at all would immediately catch. But it's a minuscule duplication. Done in ba57776b1968126154f24fb98af8f72c9387b10a